dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/tag/TagRouter.java

```diff
@@ -141,6 +141,37 @@ public URL getUrl() {
         }
     }
 
+    /**
+     * If there's no dynamic tag rule being set, use static tag in URL.
+     *
+     * A typical scenario is a Consumer using version 2.7.x calls Providers using version 2.6.x or lower,
+     * the Consumer should always respect the tag in provider URL regardless of whether a dynamic tag rule has been set to it or not.
+     *
+     * TODO, to guarantee consistent behavior of interoperability between 2.6- and 2.7+, this method should has the same logic with the TagRouter in 2.6.x.
+     *
+     * @param invokers
+     * @param url
+     * @param invocation
+     * @param <T>
+     * @return
+     */
+    private <T> List<Invoker<T>> filterUsingStaticTag(List<Invoker<T>> invokers, URL url, Invocation invocation) {
+        List<Invoker<T>> result = invokers;
+        // Dynamic param
+        String tag = StringUtils.isEmpty(invocation.getAttachment(TAG_KEY)) ? url.getParameter(TAG_KEY) :
+                invocation.getAttachment(TAG_KEY);
+        // Tag request
+        if (!StringUtils.isEmpty(tag)) {
+            result = filterInvoker(invokers, invoker -> tag.equals(invoker.getUrl().getParameter(Constants.TAG_KEY)));
+            if (CollectionUtils.isEmpty(result) && !isForceUseTag(invocation)) {
```

Comment:
Is this 
```
if (CollectionUtils.isEmpty(result) && !isForceUseTag(invocation)) {
                result = filterInvoker(invokers, invoker -> StringUtils.isEmpty(invoker.getUrl().getParameter(Constants.TAG_KEY)));
            }
```
should not be irrespective of whether **tag** present or not? I was just referring to 2.6.x implementaion where it seems to me this logic is outside of whether tag is present or not.

Reply:
Yes, I think you are right about the implementation in 2.6.x. 
But the difference here is 2.6.x does not have a deliberately defined `force` property, we can also think as it has a `force` key which is always `false`. So in 2.6.x, it will always go to the following branch at 
https://github.com/apache/incubator-dubbo/pull/3065/files#diff-179f8c2f0653380f5fd6240f6ce165b7R71
```java
f (result.isEmpty()) {
            // Only forceTag = true force match, otherwise downgrade
            String forceTag = RpcContext.getContext().getAttachment(Constants.FORCE_USE_TAG);
            if (StringUtils.isEmpty(forceTag) || "false".equals(forceTag)) {
                for (Invoker<T> invoker : invokers) {
                    if (StringUtils.isEmpty(invoker.getUrl().getParameter(Constants.TAG_KEY))) {
                        result.add(invoker);
                    }
                }
            }
        }
```
irrespective of whether tag present or not.

For 2.7, you can see that we have covered the branch in which tag is not present. 
```java
else {
            result = filterInvoker(invokers, invoker -> StringUtils.isEmpty(invoker.getUrl().getParameter(Constants.TAG_KEY)));
        }
```
But when the tag is empty, I think there's no need to check force, `force=true` only makes sense to the scenario when the tag is not empty.

Reply:
Thanks for clearing my doubts. 👍